src/cluster.py

```diff
@@ -0,0 +1,226 @@
+
+    def __init__(self, version):
+        self.version = version
+        self.conf_path = Path(f"/etc/postgresql/{version}/main")
+
+    def bootstrap_cluster(self, password: str) -> None:
+        """Bootstrap a PostgreSQL cluster with the given superuser password."""
+        # Check for no running clusters (like the default cluster created on postgres install).
+        if self._is_cluster_running():
+            raise ClusterAlreadyRunningError()
+        else:
+            # Create a new cluster.
+            self._create_cluster(password)
+            # Render the configuration files and start the cluster.
+            self._render_pg_hba_conf_file()
+            self._render_postgresql_conf_file()
+            self._start_cluster()
+            # Check that the cluster is up and running.
+            if not self._is_cluster_running():
+                raise ClusterNotRunningError()
+
+    def inhibit_default_cluster_creation(self) -> None:
+        """Stop the PostgreSQL packages from creating the default cluster."""
+        os.makedirs(os.path.dirname(CREATE_CLUSTER_CONF_PATH), mode=0o755, exist_ok=True)
+        with open(CREATE_CLUSTER_CONF_PATH, mode="w") as file:
+            file.write("create_main_cluster = false\n")
+            file.write(f"include '{self.conf_path}/conf.d/postgresql-operator.conf'")
+
+    def _create_cluster(self, password: str) -> None:
+        """Create a new PostgreSQL cluster."""
+        # Write the password in a temporary file for pg_createcluster.
+        with open("/tmp/pwfile", mode="w") as file:
+            file.write(password)
+        try:
+            command = [
+                "pg_createcluster",
+                self.version,
+                "main",
+                "--datadir=/var/lib/postgresql/data/pgdata",
+                "--",
+                "--pwfile=/tmp/pwfile",
+            ]
+            logger.info(f"pg_createcluster call: {' '.join(command)}")
+            subprocess.check_call(command)
+        except subprocess.CalledProcessError as e:
+            raise ClusterCreateError(e.stdout)
+        finally:
+            # Remove the temporary file.
+            os.remove("/tmp/pwfile")
+
+    def _is_cluster_running(self) -> bool:
+        """Check whether the cluster is running or not."""
+        try:
+            clusters = (
+                subprocess.check_output(["pg_lsclusters", "--no-header"]).decode().splitlines()
+            )
+
+            def is_online(cluster: str):
+                """Return whether or not a cluster is online."""
+                if cluster.split()[3] == "online":
+                    return True
+                return False
+
+            online_clusters = list(filter(is_online, clusters))
+            return len(online_clusters) == 1
+        except subprocess.CalledProcessError:
+            raise
+
+    def _render_file(self, path: str, content: str, mode: int) -> None:
+        """Write a content rendered from a template to a file."""
```

Comment:
Probably worth documenting the args in this docstring

Reply:
It makes sense. I've documented the args of this method.